release/dist/qnap/pkgs.go

```diff
@@ -0,0 +1,188 @@
+	"os/exec"
+	"path/filepath"
+	"sync"
+
+	"tailscale.com/release/dist"
+)
+
+type target struct {
+	goenv  map[string]string
+	arch   string
+	signer dist.Signer
+}
+
+func (t *target) String() string {
+	return fmt.Sprintf("qnap/%s", t.arch)
+}
+
+func (t *target) Build(b *dist.Build) ([]string, error) {
+	// Stop early if we don't have docker running.
+	if _, err := exec.LookPath("docker"); err != nil {
+		return nil, fmt.Errorf("docker not found, cannot build: %w", err)
+	}
+
+	qnapBuilds := getQnapBuilds(b)
+	inner, err := qnapBuilds.buildInnerPackage(b, t.goenv)
+	if err != nil {
+		return nil, err
+	}
+
+	return t.buildQPKG(b, qnapBuilds, inner)
+}
+
+func (t *target) buildQPKG(b *dist.Build, qnapBuilds *qnapBuilds, inner *innerPkg) ([]string, error) {
+	if _, err := exec.LookPath("docker"); err != nil {
+		return nil, fmt.Errorf("docker not found, cannot build: %w", err)
+	}
+
+	if err := qnapBuilds.makeDockerImage(b); err != nil {
+		return nil, fmt.Errorf("makeDockerImage: %w", err)
+	}
+
+	qnapTag := "1" // currently static, we don't seem to bump this
+
+	filename := fmt.Sprintf("Tailscale_%s-%s_%s.qpkg", b.Version.Short, qnapTag, t.arch)
+	out := filepath.Join(b.Out, filename)
+	log.Printf("Building %s", out)
+
+	cmd := b.Command(b.Repo, "docker", "run", "--rm",
+		"-e", fmt.Sprintf("ARCH=%s", t.arch),
+		"-e", fmt.Sprintf("TSTAG=%s", b.Version.Short),
+		"-e", fmt.Sprintf("QNAPTAG=%s", qnapTag),
+		"-v", fmt.Sprintf("%s:/tailscale", inner.tsPath),
+		"-v", fmt.Sprintf("%s:/tailscaled", inner.tsdPath),
+		// Tailscale folder has QNAP package setup files needed for building.
+		"-v", fmt.Sprintf("%s:/Tailscale", filepath.Join(b.Repo, "release/dist/qnap/Tailscale")),
+		"-v", fmt.Sprintf("%s:/build-qpkg.sh", filepath.Join(b.Repo, "release/dist/qnap/build-qpkg.sh")),
+		"-v", fmt.Sprintf("%s:/out", b.Out),
+		"build.tailscale.io/qdk:latest",
+		"/build-qpkg.sh",
+	)
+	// dist.Build runs target builds in parallel goroutines by default.
+	// For QNAP, this is an issue because the underlaying qbuild builder will
+	// create tmp directories in the shared docker image that end up conflicting
+	// with one another.
+	// So we use a mutex to only allow one "docker run" at a time.
+	qnapBuilds.dockerImageMu.Lock()
+	defer qnapBuilds.dockerImageMu.Unlock()
+	if err := cmd.Run(); err != nil {
+		return nil, fmt.Errorf("docker run: %w", err)
+	}
+
+	// TODO(sonia): Use signer?
+
+	return []string{out, out + ".md5"}, nil
+}
+
+type qnapBuildsMemoizeKey struct{}
+
+type innerPkg struct {
+	tsPath  string // tailscale binary path
+	tsdPath string // tailscaled binary path
+}
+
+// qnapBuilds is extra build context shared by all qnap builds.
+type qnapBuilds struct {
+	innerPkgs     dist.Memoize[*innerPkg]
+	dockerImageMu sync.Mutex
+}
+
+// getQnapBuilds returns the qnapBuilds for b, creating one if needed.
+func getQnapBuilds(b *dist.Build) *qnapBuilds {
+	return b.Extra(qnapBuildsMemoizeKey{}, func() any { return new(qnapBuilds) }).(*qnapBuilds)
+}
```

Comment:
can we create the `qnapBuilds` instance in `func Targets` and pass it as a field in `target`?
that might be a little easier to follow than this implicit memoized state

Reply:
I actually mimicked what Synology's builder was doing here. I think for consistency we should keep them aligned, but happy to follow up with cleaning both of them up to create their builders from `Targets` and pass it through.